lib/core/include/data_object_proxy.hpp

```diff
@@ -0,0 +1,212 @@
+        using doi_type = I;
+        using doi_pointer_type = doi_type*;
+        using replica_list = std::vector<replica::replica_proxy<doi_type>>;
+        using size_type = int;
+
+        /// \brief Constructs proxy using an existing doi_type
+        /// \since 4.2.9
+        explicit data_object_proxy(doi_type& _doi)
+            : doi_{&_doi}
+            , r_{fill_replica_list(_doi)}
+        {
+        }
+
+        // accessors
+
+        /// \returns rodsLong_t
+        /// \retval id of the data object from the catalog
+        /// \since 4.2.9
+        auto data_id() const -> rodsLong_t { return doi_->dataId; }
+
+        /// \returns rodsLong_t
+        /// \retval id of the collection containing this data object
+        /// \since 4.2.9
+        auto collection_id() const -> rodsLong_t { return doi_->collId; }
+
+        /// \returns std::string_view
+        /// \retval Logical path of the data object
+        /// \since 4.2.9
+        auto logical_path() const -> std::string_view { return doi_->objPath; }
+
+        /// \returns std::string_view
+        /// \retval Owner user name of the data object
+        /// \since 4.2.9
+        auto owner_user_name() const -> std::string_view { return doi_->dataOwnerName; }
+
+        /// \returns std::string_view
+        /// \retval Owner zone name of the data object
+        /// \since 4.2.9
+        auto owner_user_zone() const -> std::string_view { return doi_->dataOwnerZone; }
+
+        /// \returns size_type
+        /// \retval Number of replicas for this data object
+        /// \since 4.2.9
+        auto replica_count() const -> size_type { return r_.size(); }
+
+        /// \returns const replica_list&
+        /// \retval Reference to the list of replica_proxy objects
+        /// \since 4.2.9
+        auto replicas() const -> const replica_list& { return r_; }
+
+        /// \returns const doi_pointer_type
+        /// \retval Pointer to the underlying struct
+        /// \since 4.2.9
+        auto get() const -> const doi_pointer_type { return doi_; }
+
+        // mutators
+
+        /// \brief Set the data id for the data object
+        /// \param[in] _did - New value for data id
+        /// \returns void
+        /// \since 4.2.9
+        template<
+            typename P = doi_type,
+            typename = std::enable_if_t<!std::is_const_v<P>>>
+        auto data_id(const rodsLong_t _did) -> void
+        {
+            for (auto& r : r_) {
+                r.data_id(_did);
+            }
+        }
+
+        /// \brief Set the collection id for the data object (i.e. all replicas)
+        /// \param[in] _cid - New value for collection id
+        /// \returns void
+        /// \since 4.2.9
+        template<
+            typename P = doi_type,
+            typename = std::enable_if_t<!std::is_const_v<P>>>
+        auto collection_id(const rodsLong_t _cid) -> void 
+        {
+            // TODO: should this set the logical path too?
```

Comment:
i think it must?    what does it mean not to?

Reply:
Well, the setting of the logical path via change in collection_id is necessarily a hit to the database in order to resolve the ID to a logical path so that we can set it (unless we have a resource_manager-esque entity for collections that I do not know about). Are we willing to take that hit, or is it up to the caller (or a separate member function) to do this if desired?

Reply:
is there any upside to allowing someone to get the database 'out of sync' in this way?   i can only think of how this can be bad.

Reply:
If this does not set the logical path (or if setting the logical path does not set the collection id), one or the other is out of sync within this data model. We plan to use this structure to generate r_data_main updates which will be applied in the finalize API. As collection name is not stored in the catalog, I think this is okay. The only time this would be happening would be in a rename operation anyway, so the physical path would need to be updated for each replica as well.

After some more thinking, I feel this is the responsibility of a higher-level call, whether in the caller or in a convenience function. Thoughts?

Reply:
I agree.  This sounds like the responsibility of another API.

Is it the proxy's responsibility to keep things in sync or is it something that simply provides a better interface to the C type?

Reply:
I think if the need ever arises, a new member function can be introduced to keep everything aligned. For now, it won't be concerned with keeping things in sync to keep things simple and the caller will need to make sure that the information is correct if it is going to be updating the catalog with information from these structs.

Reply:
seems fair.

Reply:
After offline discussion, we have determined both member functions should be replaced by a single `rename` function which updates both the collection id and the logical path as provided by the caller so that the ORM cannot be out of sync.